### ISSUE TITLE
VoiceOver has 5 second delay before reading aria-describedby content

### ISSUE THREAD

**Issue by jessegreenberg** [phetsims/balloons-and-static-electricity#183]
VoiceOver has a 5 second delay before reading aria-describedby content.  Is this an issue for application style content?

For example, when the  balloon is picked up, VoiceOver will:
- Read the Balloon Label
- Wait 5 seconds
- Read a preamble stating that the user is in HTML content
- Read the description about the balloon

Will VoiceOver users wait to hear the description, or quickly tab to the next thing?  Arrow keys might not be usable to find the description when the balloon is picked up.

If an issue, we could try to use aria-live to force the description to be read immediately, and remove the aria-describedby attribute.  aria-live is now working quite consistently on all major AT/Browser combinations.

@terracoda @emily-phet @jhung @jobara do you think this would be an issue?

**Comment by jobara** [comment]
@jessegreenberg good question, I think it might require some testing with a VoiceOver user. My initial feeling is that we should avoid work arounds and hacks as much as possible. I wonder if this is the intended behaviour from Apple so that a user could wait and listen to the detail if the want/need to and to skim past it in other circumstances. Sort of like a tooltip. It would also be good to confirm this behaviour in a constrained example outside of the sim code, to ensure that there is nothing else causing this behaviour.

**Comment by terracoda** [comment]
@jessegreenberg, @jobara,  @emily-phet,  @jhung I was playing with VO setting yesterday and see that there is a setting for a delay. Mine, by default was set for a long delay. It is not clear to me if this delay is just for VO "tool-tip" type stuff, or other content. There are settings for reading the DOM and verbosity as well. It would be good to ask VO users about their settings in our next interviews.

**Comment by terracoda** [comment]
@jessegreenberg, @jobara, @emily-phet, @jhung I have noticed that the DOM setting makes it easier for me to read/navigate with the VO cursor keys.

**Comment by terracoda** [comment]
@jessegreenberg, @jobara, @emily-phet, @jhung 
> I was playing with VO setting yesterday and see that there is a setting for a delay.

I think this delay setting is just for VO hints.

**Comment by jessegreenberg** [comment]
>I think it might require some testing with a VoiceOver user...I wonder if this is the intended behaviour from Apple so that a user could wait and listen to the detail if the want/need to and to skim past it in other circumstances. Sort of like a tooltip.

Good point, I think that is likely the case.  Since the label is read immediately, maybe this is enough for VO users to understand that they have picked up the balloon.

> It would also be good to confirm this behaviour in a constrained example outside of the sim code, to ensure that there is nothing else causing this behaviour

The commit above shows another case outside of a sim.  In this case as well, the delay is 5 seconds before reading the preamble and the aria-describedby content.

@terracoda did adjusting the delay change the delay time of reading the description of the balloon?

**Comment by jessegreenberg** [comment]
I don't really see using aria-live instead of aria-describedby as a hack.  This sim already uses aria-live to present information to the user during interaction with the balloon.

**Comment by jobara** [comment]
@jessegreenberg sorry, i didn't mean to imply that aria-live was a hack. I was really being more general in that comment. It's just to say that often adhering closer to spec and best practices will provide something more robust.

**Comment by jessegreenberg** [comment]
@jobara, I see, thanks.  Yes, I definitely agree with that!  Adhering to standards would be best.

**Comment by terracoda** [comment]
@jessegreenberg, changing the delay settings didn't do anything for sim content, only VO content.

**Comment by terracoda** [comment]
@jessegreenberg I just sat with a VO user here at Assets and we think the live region content is not reading out because of lack of bandwidth. I will try a downloaded version to see if I can get that working. @jessegreenberg, do you mind posting the parameter for me again? I think it is 
> ?accessibility

**Comment by jessegreenberg** [comment]
@terracoda once the simulation is downloaded, accessibility features do not require internet access.  All accessibility strings are self contained in the application, so it can't be a bandwidth issue.

Unfortunately, the downloaded version of the sim currently suffers from https://github.com/phetsims/balloons-and-static-electricity/issues/184.  The parallel DOM is in the downloaded version twice.  If you use a downloaded version, you will have to tab through an an extra copy of the PDOM before you see the focus highlight.

**Comment by terracoda** [comment]
@jessegreenberg, thanks for that comment. Yes, the visual focus shows up on the second go around, but I still do not get any interactive alerts in the local version which seems strange.

**Comment by terracoda** [comment]
@jessegreenberg, another complication is that the grabbed state is easily lost visually by using the cursor keys to read the PDOM, but the balloon doesn't release when the user does that, and then it is difficult to grab it again because it is already grabbed.

**Comment by terracoda** [comment]
@jessegreenberg, I think you are right. The issue might be more complicated than just bandwidth.

**Comment by jessegreenberg** [comment]
> another complication is that the grabbed state is easily lost visually by using the cursor keys to read the PDOM

This is a new issue, I will create a new ticket and investigate there.

**Comment by terracoda** [comment]
@jessegreenberg, @jobara I miss understood this issue at first. I didn't realize the delay was specifically for aria-describedby, I thought is was for aria-live blocks. It is possible because aria-describedby is technically for important tooltip-like information that VO delays the reading of it because it is technically a tooltip. 
@jessegreenberg, I am assume that there is no aria-describedby on the grab button, and this is just an issue for the balloon once it is in a grabbed state (i.e. after pressing the grab button).

I'd like to discuss, using aria-live and an interaction alert here which adds an indication the balloon is grabbed... For example:

1. User presses Grab Yellow balloon (button).
2. Int. Alert: Yellow balloon, grabbed. In center of play area. Has no more negative charges than positive charges. Press W, A, S, or D keys to drag balloon up, left, down, or right.

Currently, we don't have the "Yellow balloon, grabbed", content in there.

**Comment by terracoda** [comment]
@jessegreenberg, I meant "discuss and try." 

With an aria-live interaction alert, would it be easier to take out the WASD directions, once the user learns them? At the beginning, it seems to take them a couple of read of the WASD instructions, but then once they actually do it, it seems quite easy.

**Comment by jessegreenberg** [comment]
> I assume that there is no aria-describedby on the grab button, and this is just an issue for the balloon once it is in a grabbed state (i.e. after pressing the grab button).

Yes, that is correct.

@terracoda lets discuss using aria-live at our meeting today.  Hearing 'Yellow Balloon' when the user presses the 'Grab Balloon' button will be fixed by #192.  If you would like to hear 'grabbed', it might make more sense to use aria-live since this acts more as a status alert.  Otherwise I agree with https://github.com/phetsims/balloons-and-static-electricity/issues/183#issuecomment-255735460 that it would be cleaner to use aria-describedby.  'grabbed' is not included in the interaction examples (https://docs.google.com/spreadsheets/d/1b6zfzdbRwYDoHA5ecGEp8atx4moFJOf7aYXNGDFF1bE/edit#gid=0)

**Comment by jessegreenberg** [comment]
@jobara and @jessegreenberg also tested this out on MacOS Sierra and OSX Yosemite.  The delay is caused by the tooltip speed as mentioned by @terracoda in https://github.com/phetsims/balloons-and-static-electricity/issues/183#issuecomment-257886463.  By placing the setting on 'short' the delay is removed and the preamble + description is read immediately.

On Yosemite, the delay is set to the shortest setting by default.  On Yosemite, the default delay is 5 seconds.  Since the delay is removed for the latest version of MacOS, I believe that this is no longer an issue.  We will still discuss the possibility of using aria-live for this case in BASE.

**Comment by terracoda** [comment]
@jessegreenberg and @jobara, I just tested on VO on Mavericks and I get the delay, and it is clear it is beacsue VO thinks the entire description is a tooltip. I'd like to clarify that only the custom keyboard information is the actual tooltip. The rest is o isf the information a description of the balloon. 

So, maybe it is possible to separate directions about the custom keys, so it is identifiable as a tooltip? 
So, maybe keeping aria-describedby just for the custom key information is an option. 

What do you think of this idea?

**Comment by jessegreenberg** [comment]
The default behavior for aria-describedby on all platforms (including VoiceOver with Sierra) is to read the description immediately once the element receives focus.  I understand how the balloon description is not really 'tooltip' content, but if we remove aria-describedby from the description then we will only hear the keyboard help information when the balloon gets focus.

**Comment by jessegreenberg** [comment]
Marking as high priority, this should be discussed before #188.

**Comment by jessegreenberg** [comment]
11/2/16 meeting - We decided that this should not be an issue for the next student interview.  We discussed that the crux of this issue is that aria-describedby is in general placed in a category of tooltip information, which is is not appropriate for complex interactive content.

I am leaving this issue open.  @terracoda found that classifying description content in 'tooltip' information is confusing.  After interviews, @terracoda will review this issue and determine if changes need to be made so that the description comes without the typical VoiceOver help preamble.

**Comment by jessegreenberg** [comment]
In #215 @terracoda said

> Upon grabbing the balloon, the initial description containing position, charge and, instructions rarely fires (VO in Safari). Need to test further.

It seems very likely that this is the issue that is causing this to happen.

**Comment by jessegreenberg** [comment]
I think we have come to accept this, and in general we are using custom alerts when we need things to be read at a particular time. Closing.
